File: projects/backend/src/geocode.rs
use reqwest::Client;
use serde::Deserialize;

pub async fn geocode_zip(zip: &str, country: &str) -> Result<Location, Box<dyn std::error::Error>> {
    let url = format!("https://nominatim.openstreetmap.org/search.php?country={country}&postalcode={zip}&format=jsonv2");
    let client = Client::new();
    let response = client
        .get(&url)
        .header("User-Agent", "PowerSquirrel (igel@igels.net)")
        .send()
        .await?;

    if response.status().is_success() {
        let results: Vec<NominatimResult> = response.json().await?;
        if let Some(first_result) = results.into_iter().next() {
            let lat = first_result.lat.parse::<f64>()?;
            let lon = first_result.lon.parse::<f64>()?;
            return Ok(Location { lat, lon });
        } else {
            return Err("No results found in response".into());
        }
    } else {
        return Err(format!("Request failed with status {}", response.status()).into());
    }
}

#[derive(Debug, Deserialize)]
pub struct Location {
    pub lat: f64,
    pub lon: f64,
}

#[derive(Debug, Deserialize)]
struct NominatimResult {
    lat: String,
    lon: String,
}

File: projects/backend/src/backend/geocode.clj
(ns backend.geocode
  (:require [clj-http.client :as http]
            [cheshire.core :as json]))

(defn geocode-zip [zip country]
  (let [url (format "https://nominatim.openstreetmap.org/search.php?country=%s&postalcode=%s&format=jsonv2"
                    (java.net.URLEncoder/encode country "UTF-8")
                    (java.net.URLEncoder/encode zip "UTF-8"))
        resp (http/get url {:headers {"User-Agent" "PowerSquirrel (igel@igels.net)"}
                            :accept :json
                            :as :text})]
    (when (= 200 (:status resp))
      (let [data (json/parse-string (:body resp) true)
            first (first data)]
        (when first
          {:lat (Double/parseDouble (:lat first))
           :lon (Double/parseDouble (:lon first))})))))

